### ISSUE TITLE
Individual tuning of @NonNullByDefault for record components

### ISSUE THREAD

**Issue by stephan-herrmann** [eclipse-jdt/eclipse.jdt.core#701]
I suggest we add a constant `RECORD_COMPONENT` to `DefaultLocation` such that `@NonNullByDefault` can be fine tuned separately for record components.

I was thinking if we could get away without this:
* The constant - `FIELD` looks like a good match, 
  * but `FIELD` is not a good option for some projects, because it's quite hard to ensure assignment to all fields right from the constructor. For records, however, assignment in the constructor *is* ensured, so applying `@NonNullByDefault` to record components but not to fields is a reasonable choice.
* A record component actually surfaces in a constructor parameter and an accessor return type (while the field is hidden), so we _could_ theoretically control record components by the "union" of `PARAMETER` and `RETURN_TYPE`
  * but that would be quite unexpected / difficult to communicate to users

Conversely, there is already a similar constant `java.lang.annotation.ElementType.RECORD_COMPONENT`, which suggests we should add the same in `org.eclipse.jdt.annotation.DefaultLocation`.

**Comment by stephan-herrmann** [comment]
@iloveeclipse @jarthana does this proposal make sense to you? 

If so, what version bump is needed for addition of a constant in an API enum? Note, that this enum is not _meant_ to be referenced in actual user code, just as an annotation value, so we shouldn't expect any switch statements requiring an additional case statement. I guess a minor version bump is indicated here (2.2.x -> 2.3)?

**Comment by iloveeclipse** [comment]
Correct, according to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_interfaces_-_API_fields, adding enum field is binary compatible and because it is externally visible, according to https://wiki.eclipse.org/Version_Numbering#When_to_change_the_minor_segment should trigger a minor segment bump:

> The minor segment number must be incremented when a plug-in changes in an "externally visible" way. Examples of externally visible changes include [binary compatible API changes](https://wiki.eclipse.org/Evolving_Java-based_APIs_2)

**Comment by stephan-herrmann** [comment]
> Correct, according to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_interfaces_-_API_fields, adding enum field is binary compatible

Ah, thanks. I had only looked at part 1 of that guide, and thus didn't find a conclusive answer. Silly me :)

**Comment by stephan-herrmann** [comment]
Reminder to self: In the end I want to check if all questions from https://bugs.eclipse.org/bugs/show_bug.cgi?id=571987 have been answered.

**Comment by stephan-herrmann** [comment]
> Reminder to self: In the end I want to check if all questions from https://bugs.eclipse.org/bugs/show_bug.cgi?id=571987 have been answered.

While revisiting that bug I found a few gaps still, to be fixed via #4283
